Source/WebCore/rendering/RenderLayer.cpp

```diff
@@ -3389,7 +3389,7 @@ void RenderLayer::paintList(LayerList layerIterator, GraphicsContext& context, c
     if (layerIterator.begin() == layerIterator.end())
         return;
 
-    if (!hasSelfPaintingLayerDescendant())
+    if (!hasSelfPaintingLayerDescendant() || renderer().shouldSkipContent())
```

Comment:
How is it determined where these tests are needed? It seems bit random.

The spec says "...to a large extent do not update their styles at all unless explicitly requested by script" and "user agent should additionally avoid as much layout/rendering work as possible for skipped contents" yet I see only tests about skipping painting and hit testing. This won't really help performance much. Is the plan to add those things later?

The spec also says "This is similar to giving the contents display: none". In display:none case we don't actually build a render tree at all for that subtree. Is that the intention of the spec?

Reply:
OK, I'll let @rwlbuis explain this, he does most work of this feature:)

Reply:
The tests are added in order to comply with the specification/tests or to make things more efficient, see also the per method comments in the commit message.

A lot of the performance comes from enabling all kinds of containment when content-visibility is hidden, we did that by adjusting shouldApply*Containment methods. But note paint containment mostly works through overflow: clip, so this patch optimises more by avoiding painting of contents/children. I agree for hit testing it may not be a big win but for correctness we have to implement that (whenever the contents are skipped).

Regarding display: none, there are a lot of similarities, and I would not be surprised that its problems was the starting point for content-visibility. Basically display: none changes are heavy while content-visibility changes should be instant ideally. I think the hint was a way to help the reader think about content-visibility: hidden. In the content-visibility case I think we do want to build the render tree, to be less costly when updating the property, and to make content-visibility: auto efficient as well.

Work on top, besides the obvious content-visibility: auto work, will be in the area of editing, resize/intersection observer, accessibility. From the top of my head, except maybe resize/intersection observer, I think it is more about correctness/spec compliance than performance.